compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -1539,12 +1547,44 @@ impl EncodeContext<'a, 'tcx> {
             // so we manually encode just the information that we need
             for proc_macro in &hir.krate().proc_macros {
                 let id = proc_macro.owner.local_def_index;
-                let span = self.lazy(hir.span(*proc_macro));
+                let mut name = hir.name(*proc_macro);
+                let span = hir.span(*proc_macro);
                 // Proc-macros may have attributes like `#[allow_internal_unstable]`,
                 // so downstream crates need access to them.
-                let attrs = self.lazy(hir.attrs(*proc_macro));
-                self.tables.span.set(id, span);
-                self.tables.attributes.set(id, attrs);
+                let attrs = hir.attrs(*proc_macro);
+                let macro_kind = if tcx.sess.contains_name(attrs, sym::proc_macro) {
+                    MacroKind::Bang
+                } else if tcx.sess.contains_name(attrs, sym::proc_macro_attribute) {
+                    MacroKind::Attr
+                } else if let Some(attr) = tcx.sess.find_by_name(attrs, sym::proc_macro_derive) {
+                    // This unwrap chain should have been checked by the proc-macro harness.
+                    name = attr.meta_item_list().unwrap()[0]
+                        .meta_item()
+                        .unwrap()
+                        .ident()
+                        .unwrap()
+                        .name;
+                    MacroKind::Derive
+                } else {
+                    bug!("Unknown proc-macro type for item {:?}", id);
+                };
+
+                let mut def_key = self.tcx.hir().def_key(proc_macro.owner);
+                def_key.disambiguated_data.data = DefPathData::MacroNs(name);
+
+                let def_id = DefId::local(id);
+                record!(self.tables.kind[def_id] <- EntryKind::ProcMacro(macro_kind));
+                record!(self.tables.attributes[def_id] <- attrs);
+                record!(self.tables.def_keys[def_id] <- def_key);
+                record!(self.tables.ident_span[def_id] <- span);
+                record!(self.tables.span[def_id] <- span);
+                record!(self.tables.stability[def_id] <?- stability);
+                record!(self.tables.visibility[def_id] <- ty::Visibility::Public);
+                self.tables.deprecation.unset(id);
+                self.tables.mir.unset(id);
+                self.tables.mir_abstract_consts.unset(id);
+                self.tables.promoted_mir.unset(id);
+                self.tables.unused_generic_params.unset(id);
```

Comment:
What happens if you don't write `None`s into the table?
Isn't `None` the default state of the table if you write nothing for the given id?

Reply:
(If writing `None` is not necessary, you'll be able to get rid of `<?-`, `unset`, etc.)

Reply:
Writing None allows to skip the calls to `filter(|_| !self.is_proc_macro(id))` at decoding time. At first sight, not writing None would only hide ICEs. (Not tested.)

Reply:
As an example, `mir_abstract_consts` is written conditionally for non-proc-macros.
If there's no "MIR abstract const" for the given id, then it's not written.